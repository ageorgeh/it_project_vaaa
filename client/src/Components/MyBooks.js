/* eslint-disable no-unused-vars */

import ShelfPane from './ShelfPane'
import BookPane from './BookPane'
// eslint-disable-next-line no-unused-vars
import React, { useEffect, useState } from 'react'
import { useNavigate } from 'react-router-dom'
import { auth, logout } from '../firebase-setup'
// signInWithEmailAndPassword
import { useAuthState } from 'react-firebase-hooks/auth'
import axios from 'axios'
import { RotatingLines } from 'react-loader-spinner'

function MyBooks () {
  const url = process.env.NODE_ENV === 'production' ? 'https://it-project-vaaah-dev-api.herokuapp.com' : ''
  // renavigate user to login if not logged in
  const [user, loading] = useAuthState(auth)
  const navigate = useNavigate()

  useEffect(() => {
    if (loading) return
    if (!user) return navigate('../login')
  }, [user, loading, navigate])

  const [BookData, setBookData] = useState(null) // Array of book objects sent from API
  const [shelfData, setShelfData] = useState(null) // Array of shelf objects sent from API
  const [r, setR] = useState(false) // Refresh state
  const [title, setTitle] = useState('') // Used for update title form
  const [books, setBooks] = useState([{}]) // Books array in format with mybooks page

  const getShelves = (shelfResponse) => {
    const shelves = []
    shelfResponse.forEach((shelf) => {
      shelves.push(shelf)
    })
    shelves.sort((x, y) => { return x.name === 'All Books' ? -1 : y.name === 'All Books' ? 1 : 0 })
    return shelves
  }

  // getting all books and storing
  useEffect(() => {
    if (loading) return
    if (!user) return navigate('../login')
    const fetch = async () => {
      await user.getIdToken(/* forceRefresh */ true).then(function (idToken) {
        axios.post(url + '/MyBooks', {
          currUID: user.uid
        },
        {
          headers: {
            Authorization: 'Bearer ' + idToken
          }
        })
          .then(response => {
            setR(false)
            setBookData(response.data)
          })
          .catch(error => console.error('Error: ', error))

        axios.post(url + '/MyShelves', {
          currUID: user.uid
        },
        {
          headers: {
            Authorization: 'Bearer ' + idToken
          }
        })
          .then(response => {
            setR(false)
            setShelfData(getShelves(response.data))
          })
          .catch(error => console.error('Error: ', error))
      }).catch(function (error) {
        console.log('Error', error)
      })
    }
    const addBooksToList = async () => {
      await fetch()
    }
    addBooksToList()
  }, [user, loading, navigate, r])

  const onShelfChange = (response) => {
    setShelfData(response)
  }

  const [currShelf, setCurrShelf] = useState('All Books')
  const [currShelfName, setCurrShelfName] = useState({})

  const selectShelf = (shelfKey) => {
    setCurrShelf(shelfKey)
    setCurrShelfName(shelfKey)
    // setCurrShelfName(shelves[shelfKey])
  }

  // eslint-disable-next-line no-constant-condition
  if (BookData === null || shelfData === null) {
    return (<div ><RotatingLines height="100" width="100"/></div>)
  } else {
    return (
    <>
      <div className="flex relative bg-stone-900">
          <ShelfPane onSelect={selectShelf} shelves={getShelves(shelfData)} onShelfChange={onShelfChange} />
          <BookPane
            shelves={shelfData}
            currShelf={currShelf}
            currShelfName={currShelf}
            books={BookData}
            />
      </div>
    </>
    )
  }
}
export default MyBooks
