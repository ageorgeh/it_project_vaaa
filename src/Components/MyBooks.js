import ShelfPane from './ShelfPane';
import BookPane from './BookPane';

export default function MyBooks() {
    return (
    <>
        <ShelfPane />
        <div className="ml-64 pl-12 pt-3">
            <BookPane />
        </div>
    </>
    )
}