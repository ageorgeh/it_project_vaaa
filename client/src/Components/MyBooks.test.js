import React from 'react';
import renderer from 'react-test-renderer';
import MyBooks from './MyBooks';

it('displays mybooks', () => {
  const component = renderer.create(
    <MyBooks />,
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});