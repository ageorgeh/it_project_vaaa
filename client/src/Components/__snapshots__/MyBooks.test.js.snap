// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`displays mybooks 1`] = `
<div
  className="flex relative bg-stone-900"
>
  <div
    className="sticky top-3 flex-auto w-1/6 rounded-lg bg-stone-800 h-fit mt-2 z-1 ml-2 py-2"
  >
    <a
      className="flex items-center pl-2.5 mb-2"
      href="/img/echidna.jpg"
    >
      <span
        className="self-center text-xl ml-2 font-semibold whitespace-nowrap dark:text-white"
      >
        Shelves
      </span>
    </a>
    <ul
      className="space-y-2 border-t-2 border-stone-500 pt-2"
    >
      <li>
        <div
          className="mx-2"
        >
          <button
            className="flex items-center py-2 w-full font-normal text-stone-900 rounded-lg dark:text-white hover:bg-stone-100 pl-3 dark:hover:bg-stone-700"
            onClick={[Function]}
          >
            All Books
          </button>
        </div>
      </li>
      <li>
        <div
          className="mx-2"
        >
          <button
            className="flex items-center py-2 w-full font-normal text-stone-900 rounded-lg dark:text-white hover:bg-stone-100 pl-3 dark:hover:bg-stone-700"
            onClick={[Function]}
          >
            Fiction
          </button>
        </div>
      </li>
      <li>
        <div
          className="mx-2"
        >
          <button
            className="flex items-center py-2 w-full font-normal text-stone-900 rounded-lg dark:text-white hover:bg-stone-100 pl-3 dark:hover:bg-stone-700"
            onClick={[Function]}
          >
            Non-Fiction
          </button>
        </div>
      </li>
      <li>
        <div
          className="mx-2"
        >
          <button
            className="flex items-center py-2 w-full font-normal text-stone-900 rounded-lg dark:text-white hover:bg-stone-100 pl-3 dark:hover:bg-stone-700"
            onClick={[Function]}
          >
            To-Read
          </button>
        </div>
      </li>
      <li>
        <div
          className="mx-2"
        >
          <button
            className="place-items-center py-2 w-full pl-3 text-base font-normal text-stone-900 bg-emerald-700 rounded-lg dark:text-white hover:bg-emerald-500"
          >
            <div
              className="flex justify-center w-full"
            >
              <div
                className="w-6"
              >
                <svg
                  className="w-6 h-6"
                  fill="currentColor"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    clipRule="evenodd"
                    d="M12 2.25c-5.385 0-9.75 4.365-9.75 9.75s4.365 9.75 9.75 9.75 9.75-4.365 9.75-9.75S17.385 2.25 12 2.25zM12.75 9a.75.75 0 00-1.5 0v2.25H9a.75.75 0 000 1.5h2.25V15a.75.75 0 001.5 0v-2.25H15a.75.75 0 000-1.5h-2.25V9z"
                    fillRule="evenodd"
                  />
                </svg>
              </div>
            </div>
          </button>
        </div>
      </li>
    </ul>
  </div>
  <div
    className="w-5/6 m-2 pl-5 rounded-lg"
  >
    <div
      className="flex text-center items-center pb-2 border-b-2 mb-8 border-stone-500"
    >
      <span
        className="flex text-stone-100 ml-2 text-3xl"
      >
        All Books
      </span>
    </div>
    <div
      className="flex flex-wrap"
    >
      <div
        className="relative w-1/6 h-64 place-content-center mb-1"
      >
        <div
          className="h-48 w-32 flex justify-center items-center overflow-hidden"
        >
          <img
            alt="book cover"
            className="shrink-0 min-h-full min-w-full"
            src="img/books/gatsby.jpg"
          />
        </div>
        <div
          className="w-32"
        >
          <p
            className="text-stone-100 text-sm break-words line-clamp-2"
          >
            The Great Gatsby
          </p>
        </div>
        <div
          className="w-32"
        >
          <p
            className="text-stone-300 text-xs break-words line-clamp-1"
          >
            F. Scott Fitzgerald
          </p>
        </div>
      </div>
      <div
        className="relative w-1/6 h-64 place-content-center mb-1"
      >
        <div
          className="h-48 w-32 flex justify-center items-center overflow-hidden"
        >
          <img
            alt="book cover"
            className="shrink-0 min-h-full min-w-full"
            src="img/books/1984.jpg"
          />
        </div>
        <div
          className="w-32"
        >
          <p
            className="text-stone-100 text-sm break-words line-clamp-2"
          >
            1984
          </p>
        </div>
        <div
          className="w-32"
        >
          <p
            className="text-stone-300 text-xs break-words line-clamp-1"
          >
            George Orwell
          </p>
        </div>
      </div>
      <div
        className="relative w-1/6 h-64 place-content-center mb-1"
      >
        <div
          className="h-48 w-32 flex justify-center items-center overflow-hidden"
        >
          <img
            alt="book cover"
            className="shrink-0 min-h-full min-w-full"
            src="img/books/moby.jpg"
          />
        </div>
        <div
          className="w-32"
        >
          <p
            className="text-stone-100 text-sm break-words line-clamp-2"
          >
            Moby Dick
          </p>
        </div>
        <div
          className="w-32"
        >
          <p
            className="text-stone-300 text-xs break-words line-clamp-1"
          >
            Herman Melville
          </p>
        </div>
      </div>
      <div
        className="relative w-1/6 h-64 place-content-center mb-1"
      >
        <div
          className="h-48 w-32 flex justify-center items-center overflow-hidden"
        >
          <img
            alt="book cover"
            className="shrink-0 min-h-full min-w-full"
            src="img/books/catcher.jpg"
          />
        </div>
        <div
          className="w-32"
        >
          <p
            className="text-stone-100 text-sm break-words line-clamp-2"
          >
            The Catcher in the Rye
          </p>
        </div>
        <div
          className="w-32"
        >
          <p
            className="text-stone-300 text-xs break-words line-clamp-1"
          >
            J. D. Salinger
          </p>
        </div>
      </div>
      <div
        className="relative w-1/6 h-64 place-content-center mb-1"
      >
        <div
          className="h-48 w-32 flex justify-center items-center overflow-hidden"
        >
          <img
            alt="book cover"
            className="shrink-0 min-h-full min-w-full"
            src="img/books/alice.jpg"
          />
        </div>
        <div
          className="w-32"
        >
          <p
            className="text-stone-100 text-sm break-words line-clamp-2"
          >
            Alice in Wonderland
          </p>
        </div>
        <div
          className="w-32"
        >
          <p
            className="text-stone-300 text-xs break-words line-clamp-1"
          >
            Lewis Carrol
          </p>
        </div>
      </div>
      <div
        className="relative w-1/6 h-64 place-content-center mb-1"
      >
        <div
          className="h-48 w-32 flex justify-center items-center overflow-hidden"
        >
          <img
            alt="book cover"
            className="shrink-0 min-h-full min-w-full"
            src="img/books/anna.jpg"
          />
        </div>
        <div
          className="w-32"
        >
          <p
            className="text-stone-100 text-sm break-words line-clamp-2"
          >
            Anna Karenina
          </p>
        </div>
        <div
          className="w-32"
        >
          <p
            className="text-stone-300 text-xs break-words line-clamp-1"
          >
            Leo Tolstoy
          </p>
        </div>
      </div>
      <div
        className="relative w-1/6 h-64 place-content-center mb-1"
      >
        <div
          className="h-48 w-32 flex justify-center items-center overflow-hidden"
        >
          <img
            alt="book cover"
            className="shrink-0 min-h-full min-w-full"
            src="img/books/botton.jpg"
          />
        </div>
        <div
          className="w-32"
        >
          <p
            className="text-stone-100 text-sm break-words line-clamp-2"
          >
            The Course of Love
          </p>
        </div>
        <div
          className="w-32"
        >
          <p
            className="text-stone-300 text-xs break-words line-clamp-1"
          >
            Alain de Botton
          </p>
        </div>
      </div>
      <div
        className="relative w-1/6 h-64 place-content-center mb-1"
      >
        <div
          className="h-48 w-32 flex justify-center items-center overflow-hidden"
        >
          <img
            alt="book cover"
            className="shrink-0 min-h-full min-w-full"
            src="img/books/dune.jpg"
          />
        </div>
        <div
          className="w-32"
        >
          <p
            className="text-stone-100 text-sm break-words line-clamp-2"
          >
            Dune
          </p>
        </div>
        <div
          className="w-32"
        >
          <p
            className="text-stone-300 text-xs break-words line-clamp-1"
          >
            Frank Herbert
          </p>
        </div>
      </div>
      <div
        className="relative w-1/6 h-64 place-content-center mb-1"
      >
        <div
          className="h-48 w-32 flex justify-center items-center overflow-hidden"
        >
          <img
            alt="book cover"
            className="shrink-0 min-h-full min-w-full"
            src="img/books/ea.jpg"
          />
        </div>
        <div
          className="w-32"
        >
          <p
            className="text-stone-100 text-sm break-words line-clamp-2"
          >
            How To Do Good Better
          </p>
        </div>
        <div
          className="w-32"
        >
          <p
            className="text-stone-300 text-xs break-words line-clamp-1"
          >
            William MacAskill
          </p>
        </div>
      </div>
      <div
        className="relative w-1/6 h-64 place-content-center mb-1"
      >
        <div
          className="h-48 w-32 flex justify-center items-center overflow-hidden"
        >
          <img
            alt="book cover"
            className="shrink-0 min-h-full min-w-full"
            src="img/books/fischer.jpg"
          />
        </div>
        <div
          className="w-32"
        >
          <p
            className="text-stone-100 text-sm break-words line-clamp-2"
          >
            My 60 Memorable Games
          </p>
        </div>
        <div
          className="w-32"
        >
          <p
            className="text-stone-300 text-xs break-words line-clamp-1"
          >
            Bobby Fischer
          </p>
        </div>
      </div>
      <div
        className="relative w-1/6 h-64 place-content-center mb-1"
      >
        <div
          className="h-48 w-32 flex justify-center items-center overflow-hidden"
        >
          <img
            alt="book cover"
            className="shrink-0 min-h-full min-w-full"
            src="img/books/hobbit.jpg"
          />
        </div>
        <div
          className="w-32"
        >
          <p
            className="text-stone-100 text-sm break-words line-clamp-2"
          >
            The Hobbit
          </p>
        </div>
        <div
          className="w-32"
        >
          <p
            className="text-stone-300 text-xs break-words line-clamp-1"
          >
            J.R. Tolkien
          </p>
        </div>
      </div>
      <div
        className="relative w-1/6 h-64 place-content-center mb-1"
      >
        <div
          className="h-48 w-32 flex justify-center items-center overflow-hidden"
        >
          <img
            alt="book cover"
            className="shrink-0 min-h-full min-w-full"
            src="img/books/mockingbird.jpg"
          />
        </div>
        <div
          className="w-32"
        >
          <p
            className="text-stone-100 text-sm break-words line-clamp-2"
          >
            How to Kill a Mockingbird
          </p>
        </div>
        <div
          className="w-32"
        >
          <p
            className="text-stone-300 text-xs break-words line-clamp-1"
          >
            Harper Lee
          </p>
        </div>
      </div>
      <div
        className="relative w-1/6 h-64 place-content-center mb-1"
      >
        <div
          className="h-48 w-32 flex justify-center items-center overflow-hidden"
        >
          <img
            alt="book cover"
            className="shrink-0 min-h-full min-w-full"
            src="img/books/narnia.jpg"
          />
        </div>
        <div
          className="w-32"
        >
          <p
            className="text-stone-100 text-sm break-words line-clamp-2"
          >
            The Lion, the Witch and the Wardrobe
          </p>
        </div>
        <div
          className="w-32"
        >
          <p
            className="text-stone-300 text-xs break-words line-clamp-1"
          >
            C.S. Lewis
          </p>
        </div>
      </div>
      <div
        className="relative w-1/6 h-64 place-content-center mb-1"
      >
        <div
          className="h-48 w-32 flex justify-center items-center overflow-hidden"
        >
          <img
            alt="book cover"
            className="shrink-0 min-h-full min-w-full"
            src="img/books/potter1.jpg"
          />
        </div>
        <div
          className="w-32"
        >
          <p
            className="text-stone-100 text-sm break-words line-clamp-2"
          >
            Harry Potter and the Philosopher's Stone
          </p>
        </div>
        <div
          className="w-32"
        >
          <p
            className="text-stone-300 text-xs break-words line-clamp-1"
          >
            J.K Rowling
          </p>
        </div>
      </div>
      <div
        className="relative w-1/6 h-64 place-content-center mb-1"
      >
        <div
          className="h-48 w-32 flex justify-center items-center overflow-hidden"
        >
          <img
            alt="book cover"
            className="shrink-0 min-h-full min-w-full"
            src="img/books/potter2.jpg"
          />
        </div>
        <div
          className="w-32"
        >
          <p
            className="text-stone-100 text-sm break-words line-clamp-2"
          >
            Harry Potter and the Chamber of Secrets
          </p>
        </div>
        <div
          className="w-32"
        >
          <p
            className="text-stone-300 text-xs break-words line-clamp-1"
          >
            J.K Rowling
          </p>
        </div>
      </div>
      <div
        className="relative w-1/6 h-64 place-content-center mb-1"
      >
        <div
          className="h-48 w-32 flex justify-center items-center overflow-hidden"
        >
          <img
            alt="book cover"
            className="shrink-0 min-h-full min-w-full"
            src="img/books/potter3.jpg"
          />
        </div>
        <div
          className="w-32"
        >
          <p
            className="text-stone-100 text-sm break-words line-clamp-2"
          >
            Harry Potter and the Prisoner of Azkaban
          </p>
        </div>
        <div
          className="w-32"
        >
          <p
            className="text-stone-300 text-xs break-words line-clamp-1"
          >
            J.K Rowling
          </p>
        </div>
      </div>
      <div
        className="relative w-1/6 h-64 place-content-center mb-1"
      >
        <div
          className="h-48 w-32 flex justify-center items-center overflow-hidden"
        >
          <img
            alt="book cover"
            className="shrink-0 min-h-full min-w-full"
            src="img/books/potter4.jpg"
          />
        </div>
        <div
          className="w-32"
        >
          <p
            className="text-stone-100 text-sm break-words line-clamp-2"
          >
            Harry Potter and the Goblet of Fire
          </p>
        </div>
        <div
          className="w-32"
        >
          <p
            className="text-stone-300 text-xs break-words line-clamp-1"
          >
            J.K Rowling
          </p>
        </div>
      </div>
      <div
        className="relative w-1/6 h-64 place-content-center mb-1"
      >
        <div
          className="h-48 w-32 flex justify-center items-center overflow-hidden"
        >
          <img
            alt="book cover"
            className="shrink-0 min-h-full min-w-full"
            src="img/books/prejudice.jpg"
          />
        </div>
        <div
          className="w-32"
        >
          <p
            className="text-stone-100 text-sm break-words line-clamp-2"
          >
            Pride and Prejudice
          </p>
        </div>
        <div
          className="w-32"
        >
          <p
            className="text-stone-300 text-xs break-words line-clamp-1"
          >
            Jane Austen
          </p>
        </div>
      </div>
      <div
        className="relative w-1/6 h-64 place-content-center mb-1"
      >
        <div
          className="h-48 w-32 flex justify-center items-center overflow-hidden"
        >
          <img
            alt="book cover"
            className="shrink-0 min-h-full min-w-full"
            src="img/books/sapiens.jpg"
          />
        </div>
        <div
          className="w-32"
        >
          <p
            className="text-stone-100 text-sm break-words line-clamp-2"
          >
            Sapiens
          </p>
        </div>
        <div
          className="w-32"
        >
          <p
            className="text-stone-300 text-xs break-words line-clamp-1"
          >
            Yuval Noah Harrari
          </p>
        </div>
      </div>
      <div
        className="relative w-1/6 h-64 place-content-center mb-1"
      >
        <div
          className="h-48 w-32 flex justify-center items-center overflow-hidden"
        >
          <img
            alt="book cover"
            className="shrink-0 min-h-full min-w-full"
            src="img/books/time.jpg"
          />
        </div>
        <div
          className="w-32"
        >
          <p
            className="text-stone-100 text-sm break-words line-clamp-2"
          >
            A Brief History of Time
          </p>
        </div>
        <div
          className="w-32"
        >
          <p
            className="text-stone-300 text-xs break-words line-clamp-1"
          >
            Stephen Hawking
          </p>
        </div>
      </div>
      <div
        className="relative w-1/6 h-64 place-content-center mb-1"
      >
        <div
          className="h-48 w-32 flex justify-center items-center overflow-hidden"
        >
          <img
            alt="book cover"
            className="shrink-0 min-h-full min-w-full"
            src="img/books/walk.jpg"
          />
        </div>
        <div
          className="w-32"
        >
          <p
            className="text-stone-100 text-sm break-words line-clamp-2"
          >
            A Walk in the Woods
          </p>
        </div>
        <div
          className="w-32"
        >
          <p
            className="text-stone-300 text-xs break-words line-clamp-1"
          >
            Bill Bryson
          </p>
        </div>
      </div>
      <div
        className="relative w-1/6 h-64 place-content-center mb-1"
      >
        <div
          className="h-48 w-32 flex justify-center items-center overflow-hidden"
        >
          <img
            alt="book cover"
            className="shrink-0 min-h-full min-w-full"
            src="img/books/wild.jpg"
          />
        </div>
        <div
          className="w-32"
        >
          <p
            className="text-stone-100 text-sm break-words line-clamp-2"
          >
            Into the Wild
          </p>
        </div>
        <div
          className="w-32"
        >
          <p
            className="text-stone-300 text-xs break-words line-clamp-1"
          >
            Jon Krakauer
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;
